strawberry/schema/schema.py

```diff
@@ -214,15 +214,64 @@ class Query:
             formatted_errors = "\n\n".join(f"❌ {error.message}" for error in errors)
             raise ValueError(f"Invalid Schema. Errors:\n\n{formatted_errors}")
 
-    def get_extensions(
-        self, sync: bool = False
-    ) -> List[Union[Type[SchemaExtension], SchemaExtension]]:
-        extensions = list(self.extensions)
+    def get_extensions(self, sync: bool = False) -> List[SchemaExtension]:
+        if not self._cached_extensions:
+            extensions = []
+            if self.directives:
+                extensions = [
+                    *self.extensions,
+                    DirectivesExtensionSync if sync else DirectivesExtension,
+                ]
+            extensions.extend(self.extensions)
+            self._cached_extensions = [
+                ext if isinstance(ext, SchemaExtension) else ext(execution_context=None)
+                for ext in extensions
+            ]
+        return self._cached_extensions
```

Comment:
I am actually worried that hashing `self` would consume...

Reply:
not sure I understand :D

Reply:
oh I thought that `__hash__` in schema hashes the schema as a string... ig we can do 
```py
def __hash__(self) -> int:
	id(self)
```
(it is required that `self` would be hashable for cached_prop)

Reply:
I think what @patrick91 means in here would be to do:

```python
@cached_property
def extensions(self):
    extensions = []
    ...
    return extensions
```

But having said that, there's a problem in here. You have a `sync: bool = False` argument that is not used in the cache, meaning that it would cache different stuff depending if the first call was with `sync=True` or `sync=False`

If this continues to be a method, I think the cache need to take that flag into consideration and cache different stuff

If we go the `cached_property` way (which I personally like`, this can be solved by defining both `extensions` and `sync_extensions`.

What do you think?